src/access/AccessRestriction.cpp

```diff
@@ -0,0 +1,270 @@
+    }
+
+    if (index != nullptr)
+    {
+        *index = newIndex;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestriction::DeleteEntry(size_t index, const FabricIndex fabricIndex)
+{
+    auto entry = GetEntry(fabricIndex, index);
+    if (entry == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    else
+    {
+        mFabricEntries[fabricIndex].erase(mFabricEntries[fabricIndex].begin() + static_cast<int>(index));
+
+        for (EntryListener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+        {
+            listener->OnEntryChanged(fabricIndex, index, entry, EntryListener::ChangeType::kRemoved);
+        }
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessRestriction::UpdateEntry(size_t index, const Entry & entry, const FabricIndex fabricIndex)
+{
+    auto localEntry = GetEntry(fabricIndex, index);
+    if (localEntry != nullptr)
+    {
+        *localEntry = entry;
+
+        for (EntryListener * listener = mListeners; listener != nullptr; listener = listener->mNext)
+        {
+            listener->OnEntryChanged(fabricIndex, index, localEntry, EntryListener::ChangeType::kUpdated);
+        }
+
+        return CHIP_NO_ERROR;
+    }
+
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR AccessRestriction::Check(const SubjectDescriptor & subjectDescriptor, const RequestPath & requestPath)
+{
+    ChipLogProgress(DataManagement, "AccessRestriction: action %d", static_cast<int>(requestPath.requestType));
+
+    if (requestPath.requestType == RequestType::kRequestTypeUnknown)
+    {
+        ChipLogError(DataManagement, "AccessRestriction: RequestPath type is unknown");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // wildcard event subscriptions are allowed. All other requests require an entity id
+    if (!requestPath.entityId.has_value())
+    {
+        if (requestPath.requestType == RequestType::kEventReadOrSubscribeRequest)
+        {
+            return CHIP_NO_ERROR;
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+
+    for (auto & entry : mFabricEntries[subjectDescriptor.fabricIndex])
+    {
+        if (entry->endpointNumber != requestPath.endpoint || entry->clusterId != requestPath.cluster)
+        {
+            continue;
+        }
+
+        for (auto & restriction : entry->restrictions)
+        {
+            // a missing id is a wildcard
+            bool idMatch = !restriction.id.HasValue() || restriction.id.Value() == requestPath.entityId.value();
+            if (!idMatch)
+            {
+                continue;
+            }
+
+            switch (restriction.restrictionType)
+            {
+            case Type::kAttributeAccessForbidden:
+                if (requestPath.requestType == RequestType::kAttributeReadRequest ||
+                    requestPath.requestType == RequestType::kAttributeWriteRequest)
+                {
+                    return CHIP_ERROR_ACCESS_DENIED;
```

Comment:
It would be helpful to ensure a log message gets generated in each of these "return CHIP_ERROR_ACCESS_DENIED" paths so its super clear in the device log that the denial was due to ARL

Reply:
I kept that logging in AcccessControl.  If the ARL check fails it will say it was restricted and details are printed above the call to ARL's check.

Reply:
You can have a CHIP_ERROR_ACCESS_DENIED_BY_ARL that is converted to a different log in AccessControl, and then further updated back down to CHIP_ERROR_ACCESS_DENIED. This would allow logging to be clearer on the reason, at least.